sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -0,0 +1,1305 @@
+
+    PLACEHOLDER_UNUSED(wrappedLambda_NDITEM_1DIM);
+    // TODO : Invoke invokeLambda_NDITEM_1DIM through CM's multi-threaded
+    // kernel launching with wrappedLambda_NDITEM_1DIM and dimension info
+  }
+
+  // NDItem_2DIM
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<(DIMS == 2) &&
+                          (std::is_same<ArgT, sycl::nd_item<2>>::value)>::type
+  runIterationSpace(const sycl::range<2> &LocalSize,
+                    const sycl::range<2> &GlobalSize,
+                    const sycl::id<2> &GlobalOffset) {
+    auto wrappedLambda_NDITEM_2DIM =
+        makeWrapper_NDITEM_2DIM(MKernel, LocalSize, GlobalSize, GlobalOffset);
+
+    spaceDim[0] = (uint32_t)LocalSize[0];
+    spaceDim[1] = (uint32_t)LocalSize[1];
+
+    groupDim[0] = (uint32_t)(GlobalSize[0] / LocalSize[0]);
+    groupDim[1] = (uint32_t)(GlobalSize[1] / LocalSize[1]);
+
+    PLACEHOLDER_UNUSED(wrappedLambda_NDITEM_2DIM);
+    // TODO : Invoke invokeLambda_NDITEM_2DIM through CM's multi-threaded
+    // kernel launching with wrappedLambda_NDITEM_2DIM and dimension info
+  }
+
+  // NDItem_3DIM
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<(DIMS == 3) &&
+                          (std::is_same<ArgT, sycl::nd_item<3>>::value)>::type
+  runIterationSpace(const sycl::range<3> &LocalSize,
+                    const sycl::range<3> &GlobalSize,
+                    const sycl::id<3> &GlobalOffset) {
+    auto wrappedLambda_NDITEM_3DIM =
+        makeWrapper_NDITEM_3DIM(MKernel, LocalSize, GlobalSize, GlobalOffset);
+
+    spaceDim[0] = (uint32_t)LocalSize[0];
+    spaceDim[1] = (uint32_t)LocalSize[1];
+    spaceDim[2] = (uint32_t)LocalSize[2];
+
+    groupDim[0] = (uint32_t)(GlobalSize[0] / LocalSize[0]);
+    groupDim[1] = (uint32_t)(GlobalSize[1] / LocalSize[1]);
+    groupDim[2] = (uint32_t)(GlobalSize[2] / LocalSize[2]);
+
+    PLACEHOLDER_UNUSED(wrappedLambda_NDITEM_3DIM);
+    // TODO : Invoke invokeLambda_NDITEM_3DIM through CM's multi-threaded
+    // kernel launching with wrappedLambda_NDITEM_3DIM and dimension info
+  }
+};
+
+/// Implementation for ESIMD_CPU device interface accessing ESIMD
+/// intrinsics and LibCM functionalties requred by intrinsics
+struct ESIMDDeviceInterfaceImpl : public ESIMDDeviceInterface {
+  // Function pointer type definitions
+  using BarrierPtr_t = void (*)(void);
+  using SplitBarrierPtr_t = void (*)(uint);
+  using MemoryBasePtr_t = char *(*)(int);
+  using SlmBasePtr_t = char *(*)(void);
+  using SlmSizePtr_t = void (*)(size_t);
+
+  // Function pointers
+
+  // Intrinsics
+  BarrierPtr_t grp_barrier_ptr;
+  SplitBarrierPtr_t split_barrier_ptr;
+  BarrierPtr_t fence_ptr;
+
+  // libcm functionalities used for intrinsics such as
+  // surface/buffer/slm access
+  MemoryBasePtr_t get_surface_base_ptr;
+  SlmBasePtr_t get_slm_base_ptr;
+  SlmSizePtr_t set_slm_size_ptr;
+
+  ESIMDDeviceInterfaceImpl() {
+#ifdef __GNUC__
+    // Linux
+    void *handle = dlopen("libcm.so", RTLD_LAZY);
```

Comment:
Could you please clarify where the library will be `dlclose`'d?

Reply:
`dlclose` is not called. I'll add.